### PR TITLE
ARES Death and Consequences; The Prequel

### PR DESCRIPTION
**Pull request by Ben10083**
# About the pull request

Implements some backend stuff for planned Project ARES processor/death gameplay.

- Destruction of ARES now disables the various functionalities ARES provides (additional ones will be added, but makes it so ARES cannot bioscan, cannot use radio, and cannot talk/log data)
- ARES Apollo proc and ARES alive check proc added
- ARES now plays an announcement upon death

Outside of ARES dying turning off the subsystems, there isn't any other major consequences people will see. Plan is for ARES to handle systems such as autopilot, and for that to be affected as well

ARES being destroyed also prevents it from speaking in APOLLO or over radio, but this was already a thing. Feedback to admins over this has been added

# Explain why it's good for the game

ARES dying was always meant to have consequences, this lays additional groundwork for further changes, and has ARES's death have an impact outside of making Working Joes cry.
 
# Changelog

:cl:
add: If ARES is destroyed, systems such as bioscan also shut down, this will be expanded in the future
add: ARES plays an announcement when destroyed
code: Procs to check for APOLLO processor and if ARES is alive added
refactor: ARES subsystems modified to utilize new procs
admin: Prompt to force a Marine Bioscan only shows when ARES is unable to perform the bioscan.
/:cl:

### CODE REVIEW

File: code/modules/mob/living/silicon/decoy/decoy.dm [hunk 1]
```diff
@@ -47,11 +47,22 @@
...
 	return ..()
 
+/proc/ares_final_words(){
+	//APOLLO
+	ares_apollo_talk("APOLLO sub-system shutting down. STOP CODE: 0x000000f4|CRITICAL_PROCESS_DIED")
+
+	//GENERAL CREW
+	shipwide_ai_announcement("A Problem has been detected and the [MAIN_AI_SYSTEM] system has been shutdown. \nTechnical Information: \n\n*** STOP CODE: 0x000000f4|CRITICAL_PROCESS_DIED\n\nPossible caused by: Rapid Unscheduled Disassembly\nContact an AI Service Technician for further assistance.", title = ":(", ares_logging = null)
+}
+
```
Comment (Cthulhu80) [inline code/modules/mob/living/silicon/decoy/decoy.dm:65]: shouldn't this proc be inside the ARES_procs.dm file? With all the other globally accessible ares procs.

File: code/modules/mob/living/silicon/decoy/decoy.dm [hunk 2]
```diff
@@ -47,11 +47,22 @@
...
 	var/datum/cause_data/cause_data = create_cause_data("rapid unscheduled disassembly", src, src)
 	addtimer(CALLBACK(GLOBAL_PROC, GLOBAL_PROC_REF(explosion), loc, -1, 0, 8, 12, TRUE, FALSE, 0, cause_data), 2 SECONDS)
 	return ..()
 
+/proc/ares_final_words(){
+	//APOLLO
+	ares_apollo_talk("APOLLO sub-system shutting down. STOP CODE: 0x000000f4|CRITICAL_PROCESS_DIED")
+
+	//GENERAL CREW
+	shipwide_ai_announcement("A Problem has been detected and the [MAIN_AI_SYSTEM] system has been shutdown. \nTechnical Information: \n\n*** STOP CODE: 0x000000f4|CRITICAL_PROCESS_DIED\n\nPossible caused by: Rapid Unscheduled Disassembly\nContact an AI Service Technician for further assistance.", title = ":(", ares_logging = null)
```
Comment (Cthulhu80) [inline code/modules/mob/living/silicon/decoy/decoy.dm:63]: :(, for the title is legendary.

**Review by Cthulhu80 — COMMENTED** [review]
Admins can still force a bioscan, I know you said you were going to add stuff later but maybe it should at least state ares's current status so the admins can quickly know. 

For example create a few defines in ARES.dm like
```
#define ARES_STATUS_ACTIVE "ACTIVE"
#define ARES_STATUS_INACTIVE "INACTIVE"
```

And then inside the events tab we can change it to
```
	var/current_status = ARES_STATUS_ACTIVE
	if(!ares_is_active())
		current_status = ARES_STATUS_INACTIVE
	var/choice = tgui_alert(usr, "Are you sure you want to trigger a bioscan? Ares current status is: [current_status]", "Bioscan?", list("Yes", "No"))
```
or something like that, idk. Be mindful there are a few faction bioscans, so make sure it only states the status for ares if the admin attempts to run a bioscan for marines.

File: code/game/machinery/ARES/ARES_procs.dm [hunk 3]
```diff
@@ -77,6 +77,13 @@ GLOBAL_LIST_INIT(maintenance_categories, list(
 
 
 // ------ ARES Logging Procs ------ //
+/proc/ares_is_active(){
+	for(var/mob/living/silicon/decoy/ship_ai/ai in GLOB.ai_mob_list)
+		if(ai.stat == DEAD)
+			return FALSE //ARES dead, most other systems also die with it
+	return TRUE
+}
```
Comment (Birdtalon) [inline code/game/machinery/ARES/ARES_procs.dm:85]: No curly braces required.

File: code/modules/admin/tabs/event_tab.dm [hunk 4]
```diff
@@ -570,7 +570,13 @@
 
 	if(!admin_holder || !(admin_holder.rights & R_MOD))
 		to_chat(src, "Only administrators may use this command.")
-		return
+		return FALSE
+
+	if(!ares_is_active()){
+		to_chat(usr, SPAN_WARNING("[MAIN_AI_SYSTEM] is destroyed, and cannot talk!"))
+		return FALSE
+	}
```
Comment (Birdtalon) [inline code/modules/admin/tabs/event_tab.dm:578]: Also no curly braces.

File: code/modules/admin/tabs/event_tab.dm [hunk 5]
```diff
@@ -592,13 +598,18 @@
...
```
Comment (Birdtalon) [inline code/modules/admin/tabs/event_tab.dm:606]: Also.

File: code/modules/mob/living/silicon/decoy/decoy.dm [hunk 6]
```diff
@@ -47,11 +47,22 @@
...
+}
```
Comment (Birdtalon) [inline code/modules/mob/living/silicon/decoy/decoy.dm:64]: Also

File: code/modules/mob/living/silicon/decoy/decoy.dm [hunk 7]
```diff
@@ -47,11 +47,22 @@
 /mob/living/silicon/decoy/death(cause, gibbed, deathmessage = "sparks up and falls silent...")
 	if(stat == DEAD)
 		return FALSE
+
+	//ARES sends out last messages
+	ares_final_words();
```
Comment (Birdtalon) [inline code/modules/mob/living/silicon/decoy/decoy.dm:52]: We don't need terminators. `ares_final_words()`

**Review by Birdtalon — CHANGES_REQUESTED** [review]

**Comment by Ben10083** [comment]
Suggested changes added, prompt to ask for forcing bioscan only shows if forcing may be needed (so only if ARES dead or cannot announce)

**Pull request merged**